middleware/rate_limiter.go

```diff
@@ -0,0 +1,239 @@
+		return context.JSON(http.StatusForbidden, nil)
+	},
+}
+
+/*
+RateLimiter returns a rate limiting middleware
+
+	e := echo.New()
+
+	var inMemoryStore = NewRateLimiterMemoryStore(
+		RateLimiterMemoryStoreRate(1),
+		RateLimiterMemoryStoreBurst(3),
+		RateLimiterMemoryStoreExpiresIn(3 * time.Minute),
+	)
+
+	e.GET("/rate-limited", func(c echo.Context) error {
+		return c.String(http.StatusOK, "test")
+	}, RateLimiter(inMemoryStore))
+*/
+func RateLimiter(store RateLimiterStore) echo.MiddlewareFunc {
+	config := DefaultRateLimiterConfig
+	config.Store = store
+
+	return RateLimiterWithConfig(config)
+}
+
+/*
+RateLimiterWithConfig returns a rate limiting middleware
+
+	e := echo.New()
+
+	var inMemoryStore = NewRateLimiterMemoryStore(
+		RateLimiterMemoryStoreRate(1),
+		RateLimiterMemoryStoreBurst(3),
+		RateLimiterMemoryStoreExpiresIn(3 * time.Minute),
+	)
+
+	config := RateLimiterConfig{
+		Skipper: DefaultSkipper,
+		IdentifierExtractor: func(ctx echo.Context) (string, error) {
+			id := ctx.RealIP()
+			return id, nil
+		},
+		ErrorHandler: func(context echo.Context) error {
+			return context.JSON(http.StatusTooManyRequests, nil)
+		},
+		DenyHandler: func(context echo.Context) error {
+			return context.JSON(http.StatusForbidden, nil)
+		},
+	}
+
+	e.GET("/rate-limited", func(c echo.Context) error {
+		return c.String(http.StatusOK, "test")
+	}, RateLimiterWithConfig(config))
+*/
+func RateLimiterWithConfig(config RateLimiterConfig) echo.MiddlewareFunc {
+	if config.Skipper == nil {
+		config.Skipper = DefaultRateLimiterConfig.Skipper
+	}
+	if config.IdentifierExtractor == nil {
+		config.IdentifierExtractor = DefaultRateLimiterConfig.IdentifierExtractor
+	}
+	if config.ErrorHandler == nil {
+		config.ErrorHandler = DefaultRateLimiterConfig.ErrorHandler
+	}
+	if config.DenyHandler == nil {
+		config.DenyHandler = DefaultRateLimiterConfig.DenyHandler
+	}
+	if config.Store == nil {
+		panic("Store configuration must be provided")
+	}
+	return func(next echo.HandlerFunc) echo.HandlerFunc {
+		return func(c echo.Context) error {
+			if config.Skipper(c) {
+				return next(c)
+			}
+			if config.BeforeFunc != nil {
+				config.BeforeFunc(c)
+			}
+
+			identifier, err := config.IdentifierExtractor(c)
+			if err != nil {
+				return config.DenyHandler(c)
+			}
+
+			if !config.Store.Allow(identifier) {
+				return config.ErrorHandler(c)
+			}
+			return next(c)
+		}
+	}
+}
+
+// RateLimiterMemoryStore is the built-in store implementation for RateLimiter
+type (
+	RateLimiterMemoryStore struct {
+		visitors    map[string]*Visitor
+		mutex       sync.Mutex
+		rate        rate.Limit
+		burst       int
+		expiresIn   time.Duration
+		lastCleanup time.Time
+	}
+	// Visitor signifies a unique user's limiter details
+	Visitor struct {
+		*rate.Limiter
+		lastSeen time.Time
+	}
+)
+
+// NewRateLimiterMemoryStore returns an instance of RateLimiterMemoryStore
+func NewRateLimiterMemoryStore(
+	options ...RateLimiterMemoryStoreOption,
```

Comment:
Late to the party, but why not reuse the same mechanism s for RateLimiterWithConfig.
There we also have a config struct, that allows us to set defaults and only pass required parameters.
So although this is very extensible it differs from the rest of the implementation.

I expected something like:

```go
store := RateLimiterMemoryStoreWithConfig(RateLimiterStoreConfig{rate: 5, burst: 5})
```

or

```go
store := NewRateLimiterMemoryStore(RateLimiterStoreConfig{rate: 5, burst: 5})
```

(This is a question and not q request to change)

Reply:
This pattern is what I am used to and tbh it is really attractive.. It just felt weird adding yet another struct to the whole show...

The Rob Pike recommendation from @pafuent isn't exactly convenient IMO and if you think we should use a config struct (which I also think deep down) to maintain the pattern used throughout the framework then I can switch that up real quickly!

So do you think I should switch to a struct config or maintain the Rob Pike's standard?

Reply:
Personaly I prefer it that way to keep it consistent in our codebase.
Although the concept of Rob Pike has a big advantage of evolving APIs, mixing the concepts seems a little odd.

The helper could be a minimal one (just using rate and maybe burst) like:

```go
store := NewRateLimiterStore(5)
// Will use our defaults for burst and expires and call
// RateLimiterStoreWithConfig(RateLimiterStoreConfig{rate: 5, burst: 0, expires: ...})
```